packages/graphics/src/GraphicsGeometry.js

```diff
@@ -568,41 +492,121 @@ export class GraphicsGeometry extends BatchGeometry
             return;
         }
 
-        batchPart.size = index - batchPart.start;
-        batchPart.attribSize = attrib - batchPart.attribStart;
+        const index = this.indices.length;
+        const attrib = this.points.length / 2;
+
+        batchPart.end(index, attrib);
+
         this.indicesUint16 = new Uint16Array(this.indices);
 
         // TODO make this a const..
         this.batchable = this.isBatchable();
 
         if (this.batchable)
         {
-            this.batchDirty++;
+            this.packBatches();
+        }
+        else
+        {
+            this.buildDrawCalls();
+        }
+    }
 
-            this.uvsFloat32 = new Float32Array(this.uvs);
+    /**
+     * Affinity check
+     *
+     * @param {PIXI.FillStyle | PIXI.LineStyle} styleA
+     * @param {PIXI.FillStyle | PIXI.LineSTyle} styleB
+     */
+    _compareStyles(styleA, styleB)
+    {
+        if (!styleA || !styleB)
+        {
+            return false;
+        }
 
-            // offset the indices so that it works with the batcher...
-            for (let i = 0; i < this.batches.length; i++)
-            {
-                const batch = this.batches[i];
+        if (styleA.texture.baseTexture !== styleB.texture.baseTexture)
+        {
+            return false;
+        }
 
-                for (let j = 0; j < batch.size; j++)
-                {
-                    const index = batch.start + j;
+        if (styleA.color + styleA.alpha !== styleB.color + styleB.alpha)
+        {
+            return false;
+        }
 
-                    this.indicesUint16[index] = this.indicesUint16[index] - batch.attribStart;
-                }
+        if (!!styleA.native !== !!styleB.native)
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Test geometry for batching process.
+     *
+     * @protected
+     */
+    validateBathching()
+    {
+        if (this.dirty === this.cacheDirty)
+        {
+            return false;
+        }
+
+        if (!this.graphicsData.length)
+        {
```

Comment:
just wandering if this function should be modifying batchable (`this.batchable = true;`) ? feels like it should not mutate the state?

also may be . good time to remove all the extra ifs!

```
validateBathching()
{
    if (this.dirty !== this.cacheDirty)
    {
        if (!this.graphicsData.length)
        {
            this.batchable = true;

            return false;
        }

        for (let i = 0; i < this.graphicsData.length; i++)
        {
            const data = this.graphicsData[i];
            const fill = data.fillStyle;
            const line = data.lineStyle;

            if (fill && !fill.texture.baseTexture.valid) return false;
            if (line && !line.texture.baseTexture.valid) return false;
        }

        return true;
    }

    return false;
}
```

Reply:
I placed `this.batchable=true` there in one of PR's, I think its actually fixes a bug. You think we can move it out `validateBatching`?

Reply:
I propose we dont try to remove that line for now , address it in another PR with a test.

Reply:
Happy to wait for another separate PR on top of this one. Then we can merge to dev together. But I think we may as well address this now. We could move

```
if (!this.graphicsData.length)
        {
            this.batchable = true;

            return false;
        }
```

into `updateBatches` just before the line where we call `validateBatching`?

Reply:
Done